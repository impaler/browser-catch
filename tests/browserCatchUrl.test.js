import test from 'ava'
const path = require('path')
const fixtureServer = require('./fixtures/server/server').default
const browserCatchUrl = require('../src/index').browserCatchUrl
const { stripResult } = require('./helpers/result')
const DRIVER_TYPE = 'phantomjs'

test('catching 1 error from a url', async t => {
  let serverSettings = await fixtureServer()
  let result = await browserCatchUrl(`${serverSettings.url}/error`)
  let snapshot = stripResult(result, serverSettings.port)

  t.is(result.driverType, DRIVER_TYPE)
  t.is(result.errors.length, 1)
  t.snapshot(snapshot)
})

test('throwing on an invalid url', async t => {
  const error = await t.throws(browserCatchUrl(`http://wow/invalid`))
  t.snapshot(error.message)
})

test('catching 10 errors from a url', async t => {
  let serverSettings = await fixtureServer()
  let result = await browserCatchUrl(`${serverSettings.url}/error?count=10`)
  let snapshot = stripResult(result, serverSettings.port)

  t.is(result.driverType, DRIVER_TYPE)
  t.is(result.errors.length, 10)
  t.snapshot(snapshot)
})

test('catching no errors from a valid url', async t => {
  const serverSettings = await fixtureServer()
  const result = await browserCatchUrl(`${serverSettings.url}`)
  const snapshot = stripResult(result, serverSettings.port)

  t.is(result.driverType, DRIVER_TYPE)
  t.is(result.errors.length, 0)
  t.snapshot(snapshot)
})

test('waitForExist option waits for an element to exist while catching errors', async t => {
  const serverSettings = await fixtureServer()
  const waitDuration = 5000
  const options = {
    waitForExist: `.appended-after-${waitDuration}`
  }
  const start = new Date()
  const result = await browserCatchUrl(`${serverSettings.url}/append-element-after`, options)
  const duration = new Date() - start

  t.true(duration > waitDuration)
  t.is(result.driverType, DRIVER_TYPE)
  t.is(result.errors.length, 2)

  const snapshot = stripResult(result, serverSettings.port)
  t.snapshot(snapshot)
})

test('waitForExist & waitForExistMs options to wait for a custom time for an element to exist while catching errors', async t => {
  const serverSettings = await fixtureServer()
  const waitDuration = 1000
  const options = {
    waitForExist: `.appended-after-${waitDuration}`,
    waitForExistMs: waitDuration + 500
  }
  const start = new Date()
  const result = await browserCatchUrl(`${serverSettings.url}/append-element-after?duration=${waitDuration}`, options)
  const duration = new Date() - start

  t.true(duration > waitDuration)
  t.is(result.driverType, DRIVER_TYPE)
  t.is(result.errors.length, 2)

  const snapshot = stripResult(result, serverSettings.port)
  t.snapshot(snapshot)
})

test('run option to execute a custom async script with webdriver context and options', async t => {
  const serverSettings = await fixtureServer()
  const runScriptPath = path.resolve(__dirname, './fixtures/run-script.js')
  const options = {
    run: runScriptPath
  }
  const result = await browserCatchUrl(`${serverSettings.url}/dogs`, options)

  t.is(result.driverType, DRIVER_TYPE)
  t.is(result.errors.length, 1)

  const snapshot = stripResult(result, serverSettings.port)
  t.snapshot(snapshot)
})
