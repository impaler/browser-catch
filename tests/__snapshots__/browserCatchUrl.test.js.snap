{
  "waitForExist option waits for an element to exist while catching errors": {
    "driverType": "phantomjs",
    "errors": [
      {
        "level": "SEVERE",
        "message": "error before the element is appended\n  global code (http://127.0.1.1:9999/append-element-after:8)"
      },
      {
        "level": "SEVERE",
        "message": "thrown error after the element is appended\n  (anonymous function) (http://127.0.1.1:9999/append-element-after:14)"
      }
    ],
    "options": {
      "json": null,
      "verbose": false,
      "waitForExist": ".appended-after-5000",
      "waitForExistReverse": false,
      "waitForExistMs": 10000,
      "pause": 0,
      "run": null,
      "webdriverHost": "127.0.0.1",
      "webdriverPort": 4444,
      "driverType": "phantomjs",
      "concurrent": 1
    }
  },
  "catching 1 error from a url": {
    "driverType": "phantomjs",
    "errors": [
      {
        "level": "SEVERE",
        "message": "A regular console.error\n  error (:0)"
      }
    ],
    "options": {
      "json": null,
      "verbose": false,
      "waitForExist": false,
      "waitForExistReverse": false,
      "waitForExistMs": 10000,
      "pause": 0,
      "run": null,
      "webdriverHost": "127.0.0.1",
      "webdriverPort": 4444,
      "driverType": "phantomjs",
      "concurrent": 1
    }
  },
  "catching 10 errors from a url": {
    "driverType": "phantomjs",
    "errors": [
      {
        "level": "SEVERE",
        "message": "A regular console.error\n  global code (http://127.0.1.1:9999/error?count=10:6)"
      },
      {
        "level": "SEVERE",
        "message": "A regular console.error\n  global code (http://127.0.1.1:9999/error?count=10:6)"
      },
      {
        "level": "SEVERE",
        "message": "A regular console.error\n  global code (http://127.0.1.1:9999/error?count=10:6)"
      },
      {
        "level": "SEVERE",
        "message": "A regular console.error\n  global code (http://127.0.1.1:9999/error?count=10:6)"
      },
      {
        "level": "SEVERE",
        "message": "A regular console.error\n  global code (http://127.0.1.1:9999/error?count=10:6)"
      },
      {
        "level": "SEVERE",
        "message": "A regular console.error\n  global code (http://127.0.1.1:9999/error?count=10:6)"
      },
      {
        "level": "SEVERE",
        "message": "A regular console.error\n  global code (http://127.0.1.1:9999/error?count=10:6)"
      },
      {
        "level": "SEVERE",
        "message": "A regular console.error\n  global code (http://127.0.1.1:9999/error?count=10:6)"
      },
      {
        "level": "SEVERE",
        "message": "A regular console.error\n  global code (http://127.0.1.1:9999/error?count=10:6)"
      },
      {
        "level": "SEVERE",
        "message": "A regular console.error\n  global code (http://127.0.1.1:9999/error?count=10:6)"
      }
    ],
    "options": {
      "json": null,
      "verbose": false,
      "waitForExist": false,
      "waitForExistReverse": false,
      "waitForExistMs": 10000,
      "pause": 0,
      "run": null,
      "webdriverHost": "127.0.0.1",
      "webdriverPort": 4444,
      "driverType": "phantomjs",
      "concurrent": 1
    }
  },
  "catching no errors from a valid url": {
    "driverType": "phantomjs",
    "errors": [],
    "options": {
      "json": null,
      "verbose": false,
      "waitForExist": false,
      "waitForExistReverse": false,
      "waitForExistMs": 10000,
      "pause": 0,
      "run": null,
      "webdriverHost": "127.0.0.1",
      "webdriverPort": 4444,
      "driverType": "phantomjs",
      "concurrent": 1
    }
  },
  "waitForExist & waitForExistMs options to wait for a custom time for an element to exist while catching errors": {
    "driverType": "phantomjs",
    "errors": [
      {
        "level": "SEVERE",
        "message": "error before the element is appended\n  global code (http://127.0.1.1:9999/append-element-after?duration=1000:8)"
      },
      {
        "level": "SEVERE",
        "message": "thrown error after the element is appended\n  (anonymous function) (http://127.0.1.1:9999/append-element-after?duration=1000:14)"
      }
    ],
    "options": {
      "json": null,
      "verbose": false,
      "waitForExist": ".appended-after-1000",
      "waitForExistReverse": false,
      "waitForExistMs": 1500,
      "pause": 0,
      "run": null,
      "webdriverHost": "127.0.0.1",
      "webdriverPort": 4444,
      "driverType": "phantomjs",
      "concurrent": 1
    }
  },
  "run option to execute a custom async script with webdriver context and options": {
    "driverType": "phantomjs",
    "errors": [
      {
        "level": "SEVERE",
        "message": "dogs error\n  error (:0)"
      }
    ],
    "options": {
      "json": null,
      "verbose": false,
      "waitForExist": false,
      "waitForExistReverse": false,
      "waitForExistMs": 10000,
      "pause": 0,
      "webdriverHost": "127.0.0.1",
      "webdriverPort": 4444,
      "driverType": "phantomjs",
      "concurrent": 1
    }
  },
  "throwing on an invalid url": "\nThe url \"http://wow/invalid\" did not load, currentUrl is \"about:blank\"\n"
}